### DISCUSSION TITLE
When can a plugin ask for extensions

### DISCUSSION THREAD

**Discussion by baconpaul** [free-audio/clap#111] [General]
Interesting conversation with @tim-janik tonight about a problem his host has with surge

The problem is as follows

1. Surge asks for a timer in its `::init` method
2. His host doesn't scan for extensions until after a plugin is inited

So either

1: surge is wrong and I should move my timer to activate or
2: his host is wrong and extensions should be available in `clap_plugin::init`

but  the spec is silent on this and so we wanted to raise this discussion.

@abique @robbert-vdh @Bremmers - what say you?

We can specify it in 1.0.3 also with a change to plugin.h init.

**Comment by tim-janik** [comment]
The host code for plugin setup basically does this:
```
surge = factory->create_plugin(); // plugin may not query host, allthough it'd be fine in my case
// here surge (JUCE) does *not* support get_extension
initok = surge->init(); // here it calls: host->get_extension(CLAP_EXT_TIMER_SUPPORT) && register_timer()
// early destruction demanded per spec
if (!initok) { surge->destroy(); return; }
// after successful init, the host continues to query plugin capabilities
plugin_timer_support = surge->get_extension (CLAP_EXT_TIMER_SUPPORT); 
```
And the timer implementation does this:
```
// register_timer()
if (!plugin_timer_support) return false;
id = schedule_timer(); ...;
```

In a nutshell, I was expecting plugin->init() to maybe query host extensions, but *not* use them yet, because the host has not had a chance to query the plugin for its capabilities yet. The spec doesn't say anything specific here, so far I see the following problems:

1) How's the host supposed to complete plugin specific initializations, if it's extensions are used before it can query the plugin extension capabilities?
2) Should it be allowed for the host to query for plugin extensions during host extension methods that are already called on the host while the plugin is still in init()? (could become a bit of a mess with recursions here...)
3) How does extension *use* during init() relate to the spec requirement that demands immediate plugin->destroy() if init returns false? Tricky case: `plugin_init() { ... register_timer(); return false; }` Will this leave a stale timer? Poor host having to clean that up ;-)

**Reply by baconpaul** [reply]
Yeah I think surge is mistaken here and I should move to activate. But curious if we should update the spec also

**Reply by robbert-vdh** [reply]
Ah yes this is a good point. I think `init()` is the place to query extensions, but I'm not sure if you should also be allowed to start using them there. Somehow `clap_plugin::activate()` sounds like a bit of a weird place to register a timer, but it would probably be the most sensible place to do it (and then you can unregister the timer in deactivate).

**Reply by tim-janik** [reply]
> Yeah I think surge is mistaken here and I should move to activate. But curious if we should update the spec also

Here's a sequence I think would make sense:
1) The host calls plugin->init();
2) Then plugin->init() may call all  clap_host_t functions, but shouldn't expect host extensions to be usable just yet
3) The host sees a `true` return from plugin->init()
4) Then he host queries plugin extensions. *THIS* is the earliest point where I think the overall initialization phase can be considered "completed".
5) Plugins that need to do further work after init() (e.g. for timer setup) should call clap_host_t->request_callback() from init(). After all, this function is always provided by the host.
6) In the first call to on_main_thread(), the plugin can happily mess around with everything the host provides. Note that the spec should guarantee the plugin to rely on request_callback()+on_main_thread() to work, that is *not* an optional extension mechanism after all.

**Reply by baconpaul** [reply]
I like your solution with a request callback

**Reply by baconpaul** [reply]
Right so I guess @robbert-vdh i want to make sure a main thread timer is looking out for the juce message manager mechanism as early as possible which is why I chose init. (In vstgui I do this all in gui creat destroy but juce uses its message manager for more). That’s why I like tims solution

I think we should update the documentation for init regardless

**Comment by robbert-vdh** [comment]
Yes, I think extensions can and should only be queried during `init()`. Doing it before that is not allowed, and I don't know if doing it after that point is legal or not (but it should at least not be needed since the supported extensions shouldn't change after init). I thought this documented in the spec, but it seems I'm mistaken.

**Comment by abique** [comment]
The plugin can query the host extensions in `init()`.
It is forbidden to do it before because the host do not know yet the plugin pointer.

**Reply by baconpaul** [reply]
But can it call them?

With timer the host should allow a timer to be registered even without knowing if a plugin can listen to them? That’s the order issue here

**Reply by abique** [reply]
Yes, it is safe to call into extensions in init, because the host is already initialized at this point.

**Reply by abique** [reply]
https://github.com/free-audio/clap-host/blob/main/host/plugin-host.cc#L132

**Reply by abique** [reply]
No one can blame a plugin for calling back into the host in `init()`. So I'd just do it and keep the code simple.

**Reply by tim-janik** [reply]
> Yes, it is safe to call into extensions in init, because the host is already initialized at this point.

The host side plugin extensions pointer cannot be initialized at that point. And it's unclear if the host can even recursively ask a plugin during init() if it has an extension.

> https://github.com/free-audio/clap-host/blob/main/host/plugin-host.cc#L132

That is not spec compliant, because the return value from init() is ignored.

**Reply by robbert-vdh** [reply]
> The host side plugin extensions pointer cannot be initialized at that point. And it's unclear if the host can even recursively ask a plugin during init() if it has an extension.

Why would the host need to be able to do that?

**Reply by abique** [reply]
Yes the host shouldn't ignore the return value from `init()`.
Yes the host can ask for extensions once init() have been called (and even if init did not return yet).

What is the concrete issue that you are facing?

**Reply by tim-janik** [reply]
> > The host side plugin extensions pointer cannot be initialized at that point. And it's unclear if the host can even recursively ask a plugin during init() if it has an extension.
> 
> Why would the host need to be able to do that?

In general the host may adjust behaviour depending on the facilities of a plugin (and the other way around).
In this specific case, host->register_timer() failed because it could not (yet) confirm the availability of plugin->on_timer().
That's why I'm asking if the host can only query plugin extensions *after* init() or if it may do so recursively *during* init().

**Reply by abique** [reply]
The host can just ask for the timer extension.
It needs one flag: `didFetcPluginExtensions`, and lazily do that.

**Reply by baconpaul** [reply]
but: before init is called the host can't
so there's a chicken and egg

host makes plugin
(can't ask for extensions)
host calls plugin init
(can ask for extensions)

if init calls host->registerTimer the host hasn't yet been able to ask if the plugin can handle a callback because it can't ask for extensions from the plugin before init

a few extensions have this form. timer, posix fd. ones which imply callbacks and colalboration

so either: a host needs to register a timer whether or not a plugin supports the timer api. Or we can't call register time in init. Or a host needs to defer

but that's the basic problem

**Reply by baconpaul** [reply]
like i think @tim-janik is kinda right here. If a host is going to reject timer requests from a plugin which doesn't support the timer plugin extension, the plugin can't ask in init. Or the get extensions needs to be re-entrant during init. but that has an odd impact on the return value.

**Reply by robbert-vdh** [reply]
> but: before init is called the host can't

The host either supports timers and it can register a timer for the plugin in `init()`, or it does not support timers and it cannot register the timer. That's the reason why init is separate from the creation: during `init()` the host already has access to the plugin object. VST2 sort of sidesteps this by requiring the plugin to provide a pointer to themselves in the `audioMaster()` callback.

**Reply by baconpaul** [reply]
So sure 

but the. We need to modify the timer and fd spec to say a host should register a timer independently of if the plugin supports the timer callback extension 

Because there is no way for a host to know thst pre init

or the host can call the check re entrant fashion, so init - hosttiner - does plugin support - Return tf

**Reply by abique** [reply]
Maybe a bug in Bitwig then?

**Reply by robbert-vdh** [reply]
The former sounds to me like the correct approach. The plugin needs to query the host's extensions, and the host needs to query the plugin's extensions. A situation where an extension with both a host and plugin side is implemented by only one of the two sides, then that's an anomaly, but the host and/or the plugin should be able to handle that just fine. In this case, the host would just check whether the previously retrieved timer callback extension pointer from the plugin is non-null, and then call a function on it. If the plugin for some reason requested a timer but didn't implement the plugin side of the extension then that's not the end of the world, the timer just won't do anything.

Something similar happens with Qtractor right now: the host supports plugin parameters, but the host does not yet implement the host-side of the params extension. So you can't request a flush. My plugins just print a very confused trace log message in this situation but otherwise nothing bad happens.

**Reply by baconpaul** [reply]
Well I'm not sure.

So let me explain a bit better.

Imagine if a host supports timer. If it makes its timer registration (in pseudocode)

```
registerTimer
   if (plugin supports timer callback)
       if( start timer)
         return true
  return false
```

that's a pretty reasonable implementation

but also imagine the creation path

```
p = plugin->create()
if (!plugin->init())
   plugin->destroy()
   bail;
query plugin extensions
```

you can see that if `plugin->init` calls `host->register timer` then we have three choices

1. the host asks the plugin for an exteision without knowing it is initalized
2. the host fails to register a timer because it hasn't queried plugin extensions
3. that host implementation is incorrect and a host should start a timer and independent of the plugin. But then when does it clean it up if init fails? probably before it calls destroy right?

right now it seems bitwig is in camp 3. (If you "fix" that it will break every juce plugin so don't fix it.)

but what's the right answer? It's either a fix to the spec and juce (and leave bitwig alone until we sweep all the juce plugs) or it is something else.

but this is the problem tims host had.

**Reply by robbert-vdh** [reply]
That sounds like a problem on the host's side. Like I said in my previous comment, if the plugin requests a timer using the timer extension, then the only reasonable thing to assume would be that the plugin supports the plugin-side of the timer extension, right? Why else would it request the timer? But of course, you still need to be defense, but there's no need to make this more complicated. So the init sequence becomes:

```cpp
p = plugin->create()
if (!plugin->init())
    plugin->destroy()
    bail;

plugin_timer = plugin->get_extension(CLAP_EXT_TIMER_SUPPORT)
```

Registering a timer becomes:

```cpp
register_timer:
    start timer with ID n
    return timer ID n
```

And the host's timer proc becomes:

```cpp
on_timer:
    if (timer_id == n && plugin_timer)
        plugin_timer->on_timer(n)
```

No need for potentially problematic re-entrant function calls. And again, if the plugin uses `clap_host_timer_support` to request a timer but doesn't implement `clap_plugin_timer_support`, that's simply a bug in the plugin and that will be caught immediately in development (as you'd notice your timers don't work).

**Reply by baconpaul** [reply]
right and your bail stops timers.

So then surge is correct and tim was being too defensive? he should check the plugin_timer exists on callback moment not on registration moment.

So I propose we update the timer and fd spec to say that. A host should register a time independently of if a plugin supports the timer callback.

**Reply by robbert-vdh** [reply]
I'd say so, yes. Just assume the other side (depending on whether you're a host or a plugin) is doing the right thing, but still make sure to handle the anomalous situation where that's not the case gracefully (i.e. the null pointer check in the pseudocode above).

**Reply by abique** [reply]
By the way why do you need to register a timer so quickly?
You just need one when starting the GUI right?

**Reply by baconpaul** [reply]
Juce calls it’s message loop from several places. I start it that early because that also seems to be what the vst3 port did. I could move it to activate though

**Reply by Bremmers** [reply]
MultitrackStudio is in the "2. the host fails to register a timer because it hasn't queried plugin extensions" category. Looks like Surge XT doesn't call it on Windows, so nothing bad happens because of this.

Looks like it would be easy to solve.

**Reply by baconpaul** [reply]
yeah i think i'll modify the juce wrappers to do a call later for starting this. and it is indeed only na issue on linux.

**Reply by tim-janik** [reply]
> MultitrackStudio is in the "2. the host fails to register a timer because it hasn't queried plugin extensions" category. Looks like Surge XT doesn't call it on Windows, so nothing bad happens because of this.
> 
> Looks like it would be easy to solve.

For the record, the harm caused to Surge XT when Anklang's register_timer() refused to install it, was that the Surge XT GUI was simply a black window. I.e. the GUI rendering code wasn't triggered.

**Reply by abique** [reply]
Maybe the timer can be registered in `gui->create()` maybe?

**Reply by pauldavisthefirst** [reply]
> Yes, it is safe to call into extensions in init, because the host is already initialized at this point.

The host itself is initialized, but as has been noted, it has not had a change to query the plugin yet, which may alter its behavior w.r.t. extensions. 

I prefer the proposal mentioned above that init() can only query extensions; they can be used in activate(). That has clearer semantics for both parties.

**Reply by baconpaul** [reply]
yeah i think i agree with you @pauldavisthefirst

**Reply by baconpaul** [reply]
https://github.com/free-audio/clap-juce-extensions/pull/68 modifies the juce extensions appropriately

I'm a bit out of gas. Can someone else toss in a PR to the header file to add this comment if we all agree? I think it should say "plugins can check if a host implements an extension in init, but should not call those extensions".

If not I can get to it later tomorrow (have some travel in day tomorrow).

**Reply by abique** [reply]
The plugin can call back into the host during `init()`. It is forbidden in `create()` because the host doesn't know yet the plugin pointer.
The plugin may want to check the track-info for example in order to select an appropriate audio ports configuration.

**Reply by Bremmers** [reply]
Yeah, I was going to say something like that too. Some extensions may actually need to be used in init(), so you probably can't say "using extensions in init() isn't allowed".

**Reply by pauldavisthefirst** [reply]
Well, then this also requires a rule for extensions. One rule would be: no extension is permitted whose presence (or absence) changes the behavior of any other extension. A simpler one, similar to what @tim-janik suggested, is that the plugin cannot use the any extension in init() until all of it has finished registering them all. That would probably be a good rule anyway. However, it doesn't handle the case where host behavior in response to a plugin callback changes *after* the host scans the plugin.

**Reply by baconpaul** [reply]
Right the “one way” extensions (host do this) are a bit different than the two way extensions (host do this to me because I implement this plugin extension). Timer is in the second category

I backed down that change by the way until bws43b9 due to a bug it tickled on macos

**Reply by baconpaul** [reply]
Which Alex has fixed

**Comment by Bremmers** [comment]
I've changed my code to make registering a timer in init() work. Basically I just removed a defensive check.
timer-support.h should say something about this I think.

**Comment by tim-janik** [comment]
About the "odd" situation of the host having to cleanup resources due to extension use before init() returns false...

I think it'd be good if the spec at least recommended that plugin->init() does return false; as early as possible, best *before* it starts to *use* any host extensions.
The rationale here is to avoid hard to debug scenarios. It's not impossible to cleanup of course, but can be laborious and thus hard to get right, especially if the host coding style has to avoid being overly defensive — which started this whole issue ;-)
